### PR TITLE
[FLINK-23560][runtime] Calculate absoluteTimeMillis only once for idle metric and throughput calculation.

### PR DESCRIPTION
**Pull request by akalash**
## What is the purpose of the change

*PR should fix the performance drop after FLINK-23452 but the result is complicated. When I apply these changes right after the problematic commit I defenetelly see the improvement:

- http://codespeed.dak8s.net:8080/job/flink-benchmark-request/424/
- http://codespeed.dak8s.net:8080/job/flink-benchmark-request/425/
- http://codespeed.dak8s.net:8080/job/flink-benchmark-request/427/
But when I apply these changes on the last master(current branch). The improvement is not so obvious:

- http://codespeed.dak8s.net:8080/job/flink-benchmark-request/422/
- http://codespeed.dak8s.net:8080/job/flink-benchmark-request/423/
- http://codespeed.dak8s.net:8080/job/flink-benchmark-request/426/

I still think that this PR makes sense but perhaps we should continue our investigation at least for the sortedMultiInput
*

## Brief change log
  - *TimerGauge can receive the current absolute time millis from outside rather than calculate it by itself*
  - *ThroughputCalculator can receive the current absolute time millis from outside rather than calculate it by itself*

## Verifying this change

There are refactoring only.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by akalash** [comment]
Benchmark results for different cases:
Result before degradation: sortedMultiInput(~2145), sortedTwoInput(~2000)
Current result in master: sortedMultiInput(~2050), sortedTwoInput(~1940)
Fresh master with reverted problematic commit(my target): sortedMultiInput(~2130), sortedTwoInput(~2030)
These changes applied to master: sortedMultiInput(~2120), sortedTwoInput(~2020)

Optimization(FLINK-23648) applied right before degradation: sortedMultiInput(~2145), sortedTwoInput(~2050)
Optimization(FLINK-23648) applied right after degradation + these fix: sortedMultiInput(~2130), sortedTwoInput(~2040)

My conclusion: This fix fixes the major drop from FLINK-23452 but anyway it doesn't fully fix it. Also, we have another small problem which was introduced after FLINK-23452 and also led to the small drop. But in general, I think it is ok if we will take this fix.

File: flink-runtime/src/main/java/org/apache/flink/runtime/metrics/TimerGauge.java [hunk 1]
```diff
@@ -59,6 +59,27 @@ public synchronized void markEnd() {
...
 
+    /**
+     * Duplicate of {@link #markStart()} with ability passing the time from outside for possible
+     * optimization on calling {@link Clock#absoluteTimeMillis()}.
+     */
+    public synchronized void markStart(long absoluteTimeMillis) {
+        if (currentMeasurementStart == 0) {
+            currentMeasurementStart = absoluteTimeMillis;
+        }
+    }
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/metrics/TimerGauge.java:70]: nit:
1. make this `private void markStartUnsafe(long absoluteTimeMillis);`
2. use this unsafe method in both `public synchronized void markEnd()` and `public synchronized void markStart(long absoluteTimeMillis)` to deduplicate the code?

ditto for `markEnd`?

**Review by pnowojski — COMMENTED** [review]
Almost LGTM

File: flink-runtime/src/main/java/org/apache/flink/runtime/throughput/ThroughputCalculator.java [hunk 2]
```diff
@@ -36,36 +36,49 @@ public ThroughputCalculator(Clock clock, int numberOfSamples) {
     }
 
     public void incomingDataSize(long receivedDataSize) {
-        resumeMeasurement();
+        // Force resuming measurement.
+        if (measurementStartTime == NOT_TRACKED) {
+            measurementStartTime = clock.absoluteTimeMillis();
+        }
```
[the remark this answers is not in the document]
Comment (akalash) [reply flink-runtime/src/main/java/org/apache/flink/runtime/throughput/ThroughputCalculator.java:42]: in 99% the condition `measurementStartTime == NOT_TRACKED` would be false so it doesn't make sense to calculate the absolute time millis and then doesn't use it.

File: flink-runtime/src/main/java/org/apache/flink/runtime/metrics/TimerGauge.java [hunk 3]
```diff
@@ -47,33 +47,36 @@ public TimerGauge(Clock clock) {
...
-        }
+        markStartUnsafe(clock.absoluteTimeMillis());
     }
 
     public synchronized void markEnd() {
-        if (currentMeasurementStart != 0) {
-            currentCount += clock.absoluteTimeMillis() - currentMeasurementStart;
-            currentMeasurementStart = 0;
-        }
+        markEndUnsafe(clock.absoluteTimeMillis());
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/metrics/TimerGauge.java]: Maybe revert those changes after all for the same reason as:
https://github.com/apache/flink/pull/16772#discussion_r688401559
?

**Review by pnowojski — APPROVED** [review]
Can you squash the commits and if I don't manage to merge this, ask Yuan to merge?

**Comment by curcur** [comment]
merge per @pnowojski 's request.

**Pull request merged**